<a href="https://colab.research.google.com/github/mclellanw00/module2_intro_to_pandas/blob/main/2_1_introduction_to_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Pandas

## Pandas provides Python data frames

* Popular and established
* Inspired by R dataframes
* Built on `numpy` for fast computation

In [1]:
import pandas as pd

## Our first dataframe

In [2]:
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "Love_of_R": [2, 5, 11],
                   "years_at_wsu": [4, 17, 5]})
df.head()

Names  Python_mastery  Love_of_R  years_at_wsu
0  Iverson            10.0          2             4
1   Malone             5.0          5            17
2   Bergen             1.0         11             5

## Reading from a csv

* Most data sets will be read in from a csv or JSON data file
* `Pandas` provides `read_csv` and `read_json`

### Open a local file w/ relative path

In [3]:
# Won't work in colab
artists = pd.read_csv('./data/Artists.csv')
artists.head()

FileNotFoundError: ignored

### Open a web address

In [4]:
url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.csv"
artists =  pd.read_csv(url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992       NaN          NaN  
1       1936        0       NaN          NaN  
2       1941        0       NaN          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0       NaN          NaN

# JSON data file

* Another (more modern) storage
* Here the data is stored in row `dict`

```{json}
[
{
  "ConstituentID": 1,
  "DisplayName": "Robert Arneson",
  "ArtistBio": "American, 1930–1992",
  "Nationality": "American",
  "Gender": "Male",
  "BeginDate": 1930,
  "EndDate": 1992,
  "Wiki QID": null,
  "ULAN": null
},
{
  "ConstituentID": 2,
  "DisplayName": "Doroteo Arnaiz",
  "ArtistBio": "Spanish, born 1936",
  "Nationality": "Spanish",
  "Gender": "Male",
  "BeginDate": 1936,
  "EndDate": 0,
  "Wiki QID": null,
  "ULAN": null
},
...
```

## `pandas` can read `json` data

In [5]:
# Won't work in colab
artists =  pd.read_json('./data/Artists.json')
artists.head()

ValueError: ignored

In [6]:
json_url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.json"
artists =  pd.read_json(json_url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

## <font color="red"> Exercise 2.1.2 </font>
    
Use tab-completion and `help` to discover and explore two more methods of reading a file into a `Pandas` dataframe.


In [13]:
pd.read_excel

<function pandas.io.excel._base.read_excel(io, sheet_name=0, header=0, names=None, index_col=None, usecols=None, squeeze=False, dtype: 'DtypeArg | None' = None, engine=None, converters=None, true_values=None, false_values=None, skiprows=None, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, parse_dates=False, date_parser=None, thousands=None, comment=None, skipfooter=0, convert_float=None, mangle_dupe_cols=True, storage_options: 'StorageOptions' = None)>

In [14]:
pd.read_feather

<function pandas.io.feather_format.read_feather(path, columns=None, use_threads: bool = True, storage_options: Union[Dict[str, Any], NoneType] = None)>

In [15]:
help(pd.read_feather)

Help on function read_feather in module pandas.io.feather_format:

read_feather(path, columns=None, use_threads: bool = True, storage_options: Union[Dict[str, Any], NoneType] = None)
    Load a feather-format object from the file path.
    
    Parameters
    ----------
    path : str, path object or file-like object
        Any valid string path is acceptable. The string could be a URL. Valid
        URL schemes include http, ftp, s3, and file. For file URLs, a host is
        expected. A local file could be:
        ``file://localhost/path/to/table.feather``.
    
        If you want to pass in a path object, pandas accepts any
        ``os.PathLike``.
    
        By file-like object, we refer to objects with a ``read()`` method,
        such as a file handle (e.g. via builtin ``open`` function)
        or ``StringIO``.
    columns : sequence, default None
        If not provided, all columns are read.
    use_threads : bool, default True
        Whether to parallelize reading using

In [12]:
help(pd.read_excel)

Help on function read_excel in module pandas.io.excel._base:

read_excel(io, sheet_name=0, header=0, names=None, index_col=None, usecols=None, squeeze=False, dtype: 'DtypeArg | None' = None, engine=None, converters=None, true_values=None, false_values=None, skiprows=None, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, parse_dates=False, date_parser=None, thousands=None, comment=None, skipfooter=0, convert_float=None, mangle_dupe_cols=True, storage_options: 'StorageOptions' = None)
    Read an Excel file into a pandas DataFrame.
    
    Supports `xls`, `xlsx`, `xlsm`, `xlsb`, `odf`, `ods` and `odt` file extensions
    read from a local filesystem or URL. Supports an option to read
    a single sheet or a list of sheets.
    
    Parameters
    ----------
    io : str, bytes, ExcelFile, xlrd.Book, path object, or file-like object
        Any valid string path is acceptable. The string could be a URL. Valid
        URL schemes include http, ftp, s3, and 

> Discuss what you found here

## <font color="red"> Exercise 2.1.2 </font>
    
Read in the `Artwork.csv` from [https://github.com/MuseumofModernArt/collection](https://github.com/MuseumofModernArt/collection) and display the head of the resulting dataframe.


In [18]:
# Your code here
artwork_csv = "https://media.githubusercontent.com/media/MuseumofModernArt/collection/master/Artworks.csv"
artwork = pd.read_csv(artwork_csv)
artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

                                        ThumbnailURL Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                NaN   

  Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  Width (cm)  \
0        NaN           NaN     48.6000         NaN         NaN    168.9000   
1        NaN           NaN     40.6401         NaN         NaN     29.8451   
2        NaN           NaN     34.3000         NaN         NaN     31.8000   
3        NaN           NaN     50.8000         NaN         NaN     50.8000   
4        NaN           NaN     38.4000         NaN         NaN     19.1000   

  Seat Height (cm) Duration (sec.)  
0              NaN             NaN  
1              NaN             NaN  
2              NaN             NaN  
3              NaN             NaN  
4              NaN             NaN  

[5 rows x 29 columns]

## So what is a `DateFrame`

* Like R, Pandas focuses on columns
* Think `dict` of `(str, Series)` pairs 
* A series is a typed list-like structure

In [19]:
# This is how I imagine a dataframe
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "years_at_wsu": [4.5, 17.5, 5.5]})

In [21]:
type(df)

pandas.core.frame.DataFrame

In [20]:
type(df)

pandas.core.frame.DataFrame

## Columns are `Series` and hold one type of data

In [22]:
type(artists.BeginDate), type(artists.DisplayName)

(pandas.core.series.Series, pandas.core.series.Series)

In [23]:
artists.BeginDate.dtype, artists.DisplayName.dtype

(dtype('int64'), dtype('O'))

## Two ways to access a column

* **Method 1:** like a dictionary
    * `df["column_name"]`
* **Method 2:** like an object attribute
    * `df.column_name`
    * Only for proper names!

In [24]:
artists.BeginDate.head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

In [25]:
artists['BeginDate'].head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

## More on data types

* See all data types with `df.dtypes`
* You can set the `dtypes` when you read a dataframe
* Read more about types: [Pandas docs](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dtypes.html)

In [26]:
artists.dtypes

ConstituentID      int64
DisplayName       object
ArtistBio         object
Nationality       object
Gender            object
BeginDate          int64
EndDate            int64
Wiki QID          object
ULAN             float64
dtype: object

In [27]:
artwork.dtypes

Title                  object
Artist                 object
ConstituentID          object
ArtistBio              object
Nationality            object
BeginDate              object
EndDate                object
Gender                 object
Date                   object
Medium                 object
Dimensions             object
CreditLine             object
AccessionNumber        object
Classification         object
Department             object
DateAcquired           object
Cataloged              object
ObjectID                int64
URL                    object
ThumbnailURL           object
Circumference (cm)    float64
Depth (cm)            float64
Diameter (cm)         float64
Height (cm)           float64
Length (cm)           float64
Weight (kg)           float64
Width (cm)            float64
Seat Height (cm)      float64
Duration (sec.)       float64
dtype: object

## Setting `dtypes` with `read_csv`

We can pass a `dict` of types to `dtype` keyword

In [32]:
import numpy as np
artist_types = {'ConstituentID': np.int64,
                'DisplayName': str,
                'ArtistBio': str,
                'Nationality': str,
                'Gender':str,
                'BeginDate': np.int64,
                'EndDate': np.int64,
                'Wiki QID': str,
                'ULAN':pd.Int64Dtype()} # If you get an error ==> update pandas (see below)
artists2 = pd.read_csv('https://raw.githubusercontent.com/mclellanw00/module2_intro_to_pandas/main/data/Artists.csv', dtype = artist_types)
artists2.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID       ULAN  
0       1930     1992       NaN       <NA>  
1       1936        0       NaN       <NA>  
2       1941        0       NaN       <NA>  
3       1946        0  Q1063584  500027998  
4       1941        0       NaN       <NA>

## What's up with `ULAN` ?

* Currently, `numpy` $\rightarrow$ no missing `int`s
* Pandas correct this with the `pd.Int64Dtype()` type
    * Only available in `pandas >= 0.24.0`

In [29]:
pd.__version__

'1.3.5'

## An `Int` by any other name ...

* `np.int64` $\rightarrow$ no missing values
* `pd.Int64Dtype()` $\rightarrow$ allows `NaN`

In [33]:
artists2.dtypes

ConstituentID     int64
DisplayName      object
ArtistBio        object
Nationality      object
Gender           object
BeginDate         int64
EndDate           int64
Wiki QID         object
ULAN              Int64
dtype: object

## Preview of coming attractions

* Now we can switch `BeginDate` and `EndDate` from `0` to `np.NaN`
* We will do this in the next section

# Getting to know your data

## Basic inspection tools

* `df.head()`        first five rows
* `df.tail()`        last five rows
* `df.sample(5)`     random sample of rows
* `df.shape`         number of rows/columns in a tuple
* `df.describe()`    calculates measures of central tendency
* `df.info()`

## <font color="red"> Exercise 1: Inspect the artwork from MoMA </font>

#### Read the csv and inspect the `head`

In [34]:
artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

                                        ThumbnailURL Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                NaN   

  Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  Width (cm)  \
0        NaN           NaN     48.6000         NaN         NaN    168.9000   
1        NaN           NaN     40.6401         NaN         NaN     29.8451   
2        NaN           NaN     34.3000         NaN         NaN     31.8000   
3        NaN           NaN     50.8000         NaN         NaN     50.8000   
4        NaN           NaN     38.4000         NaN         NaN     19.1000   

  Seat Height (cm) Duration (sec.)  
0              NaN             NaN  
1              NaN             NaN  
2              NaN             NaN  
3              NaN             NaN  
4              NaN             NaN  

[5 rows x 29 columns]

In [40]:
artwork.sample(5)

Title  \
21561   THE CART STUCK IN THE MUD (plate 70, 1st suppl...   
49820   Untitled from the series Our Lives and Our Chi...   
134093  Panel from Let's Take Back Our Space: 'Female'...   
119185         Residencia Alfonso Contreras (Perspective)   
55988   TV-Brown Tint, Test Pattern 2 (TV-Braunfarben,...   

                            Artist ConstituentID  \
21561                 Marc Chagall          1055   
49820                 Robert Adams            66   
134093                Marianne Wex         74481   
119185  José Antonio Attolini Lack         43916   
55988                K. P. Brehmer           760   

                                ArtistBio Nationality BeginDate EndDate  \
21561   (French, born Belarus. 1887–1985)    (French)    (1887)  (1985)   
49820               (American, born 1937)  (American)    (1937)     (0)   
134093                (German, 1937–2020)    (German)    (1937)  (2020)   
119185               (Mexican, 1931–2012)   (Mexican)    (1931)  (2012)   
55988                 (German, 1938–1997)    (German)    (1938)  (1997)   

          Gender                  Date  \
21561     (Male)             1927-1952   
49820     (Male)               1978-81   
134093  (Female)                  1977   
119185    (Male)                c.1960   
55988     (Male)  1970, published 1972   

                                                   Medium  ...  \
21561   Etching and engraving over photogravure, print...  ...   
49820                                Gelatin silver print  ...   
134093      Gelatin silver prints, paper and ink on board  ...   
119185                            Pencil and ink on paper  ...   
55988   Stereotype and linoleum cut from a portfolio o...  ...   

                                             ThumbnailURL Circumference (cm)  \
21561                                                 NaN                NaN   
49820   http://www.moma.org/media/W1siZiIsIjEzOTg2OCJd...                NaN   
134093  http://www.moma.org/media/W1siZiIsIjQ5ODQ0MCJd...                NaN   
119185                                                NaN                NaN   
55988   http://www.moma.org/media/W1siZiIsIjIwNTY0MSJd...                NaN   

       Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  \
21561         NaN           NaN        29.6         NaN         NaN   
49820         NaN           NaN        17.7         NaN         NaN   
134093        NaN           NaN         NaN         NaN         NaN   
119185        NaN           NaN        38.0         NaN         NaN   
55988         NaN           NaN        28.7         NaN         NaN   

        Width (cm) Seat Height (cm) Duration (sec.)  
21561         24.1              NaN             NaN  
49820         17.0              NaN             NaN  
134093         NaN              NaN             NaN  
119185        90.5              NaN             NaN  
55988         37.1              NaN             NaN  

[5 rows x 29 columns]

**Task:** Write a few sentences describing an problems

There are many NaN values within the data frame as they are characterized as float64 numbers vs integers... i think XD 

#### Inspect the column names with the `columns` attribute

In [35]:
artwork.columns

Index(['Title', 'Artist', 'ConstituentID', 'ArtistBio', 'Nationality',
       'BeginDate', 'EndDate', 'Gender', 'Date', 'Medium', 'Dimensions',
       'CreditLine', 'AccessionNumber', 'Classification', 'Department',
       'DateAcquired', 'Cataloged', 'ObjectID', 'URL', 'ThumbnailURL',
       'Circumference (cm)', 'Depth (cm)', 'Diameter (cm)', 'Height (cm)',
       'Length (cm)', 'Weight (kg)', 'Width (cm)', 'Seat Height (cm)',
       'Duration (sec.)'],
      dtype='object')

**Question:** See any problems?

*Your thoughts here*

#### Inspect the tail

In [37]:
artwork.tail()

Title           Artist ConstituentID  \
139932  48 Jugendstil postcards  Various Artists          6104   
139933  48 Jugendstil postcards  Various Artists          6104   
139934  48 Jugendstil postcards  Various Artists          6104   
139935  48 Jugendstil postcards  Various Artists          6104   
139936                 Untitled              NaN           NaN   

                    ArtistBio            Nationality BeginDate EndDate Gender  \
139932  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139933  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139934  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139935  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139936                    NaN                    NaN       NaN     NaN    NaN   

           Date                           Medium  ... ThumbnailURL  \
139932  c. 1898                              NaN  ...          NaN   
139933  c. 1898                              NaN  ...          NaN   
139934  c. 1898                              NaN  ...          NaN   
139935  c. 1898                              NaN  ...          NaN   
139936     1993  Chromogenic print, printed 2012  ...          NaN   

       Circumference (cm) Depth (cm) Diameter (cm) Height (cm) Length (cm)  \
139932                NaN        NaN           NaN         9.0         NaN   
139933                NaN        NaN           NaN         9.0         NaN   
139934                NaN        NaN           NaN         9.0         NaN   
139935                NaN        NaN           NaN         9.0         NaN   
139936                NaN        NaN           NaN         NaN         NaN   

       Weight (kg)  Width (cm) Seat Height (cm) Duration (sec.)  
139932         NaN        14.0              NaN             NaN  
139933         NaN        14.0              NaN             NaN  
139934         NaN        14.0              NaN             NaN  
139935         NaN        14.0              NaN             NaN  
139936         NaN         NaN              NaN             NaN  

[5 rows x 29 columns]

#### Check out the `shape`

In [38]:
artwork.shape

(139937, 29)

**Question:** What do these number mean?

The numbers mean how ever many unique/specific values of information by the number columns in the dataframe

139937 rows and 29 columns

#### Use `describe` to compute statistics

In [36]:
artwork.describe()

ObjectID  Circumference (cm)    Depth (cm)  Diameter (cm)  \
count  139937.000000           10.000000  15166.000000    1464.000000   
mean   102094.204656           44.868020     15.227029      23.226111   
std     90877.170308           28.631604     52.885909      44.902258   
min         2.000000            9.900000      0.000000       0.635000   
25%     37161.000000           23.500000      0.000000       7.778800   
50%     74919.000000           36.000000      0.000000      13.652500   
75%    144644.000000           71.125000      7.800000      25.000000   
max    436925.000000           83.800000   1808.483617     914.400000   

         Height (cm)  Length (cm)    Weight (kg)     Width (cm)  \
count  122889.000000   742.000000     287.000000  121968.000000   
mean       37.420501    89.687579    1299.039319      37.935016   
std        49.478379   329.428165   12079.491475      67.180817   
min         0.000000     0.000000       0.090000       0.000000   
25%        17.800000    17.100000       5.896800      17.500000   
50%        27.800000    26.700000      20.865500      25.400000   
75%        43.815088    79.700000      80.967100      44.132588   
max      9140.000000  8321.056600  185067.585957    9144.000000   

       Seat Height (cm)  Duration (sec.)  
count               0.0     1.949000e+03  
mean                NaN     6.081282e+03  
std                 NaN     1.434418e+05  
min                 NaN     0.000000e+00  
25%                 NaN     1.200000e+02  
50%                 NaN     4.200000e+02  
75%                 NaN     1.500000e+03  
max                 NaN     6.283065e+06

#### Use `info` to look at types and totals

In [39]:
artwork.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139937 entries, 0 to 139936
Data columns (total 29 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Title               139898 non-null  object 
 1   Artist              138711 non-null  object 
 2   ConstituentID       138711 non-null  object 
 3   ArtistBio           134091 non-null  object 
 4   Nationality         138711 non-null  object 
 5   BeginDate           138711 non-null  object 
 6   EndDate             138711 non-null  object 
 7   Gender              138711 non-null  object 
 8   Date                137812 non-null  object 
 9   Medium              130213 non-null  object 
 10  Dimensions          130606 non-null  object 
 11  CreditLine          138084 non-null  object 
 12  AccessionNumber     139937 non-null  object 
 13  Classification      139936 non-null  object 
 14  Department          139937 non-null  object 
 15  DateAcquired        133251 non-nul

**Question:** What did you learn from the last two cells?

the .describe funtion outputs the basic stats information such as the count, mean, max, min, std dev., etc

the .info function lets you know how many of the objects within each of the DF are counted as non-null values as well as their dtype
